pkg/controlplane/apiserver/server.go

```diff
@@ -0,0 +1,285 @@
+// that can be run to start serving the APIs.
+type Server struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+
+	APIResourceConfigSource   serverstorage.APIResourceConfigSource
+	RESTOptionsGetter         genericregistry.RESTOptionsGetter
+	ClusterAuthenticationInfo clusterauthenticationtrust.ClusterAuthenticationInfo
+	VersionedInformers        clientgoinformers.SharedInformerFactory
+}
+
+// New returns a new instance of Master from the given config.
+// Certain config fields will be set to a default value if unset.
+// Certain config fields must be specified, including:
+// KubeletClientConfig
+func (c completedConfig) New(name string, delegationTarget genericapiserver.DelegationTarget) (*Server, error) {
+	generic, err := c.Generic.New(name, delegationTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.EnableLogsSupport {
+		routes.Logs{}.Install(generic.Handler.GoRestfulContainer)
+	}
+
+	// Metadata and keys are expected to only change across restarts at present,
+	// so we just marshal immediately and serve the cached JSON bytes.
+	md, err := serviceaccount.NewOpenIDMetadata(
+		c.ServiceAccountIssuerURL,
+		c.ServiceAccountJWKSURI,
+		c.Generic.ExternalAddress,
+		c.ServiceAccountPublicKeys,
+	)
+	if err != nil {
+		// If there was an error, skip installing the endpoints and log the
+		// error, but continue on. We don't return the error because the
+		// metadata responses require additional, backwards incompatible
+		// validation of command-line options.
+		msg := fmt.Sprintf("Could not construct pre-rendered responses for"+
+			" ServiceAccountIssuerDiscovery endpoints. Endpoints will not be"+
+			" enabled. Error: %v", err)
+		if c.ServiceAccountIssuerURL != "" {
+			// The user likely expects this feature to be enabled if issuer URL is
+			// set and the feature gate is enabled. In the future, if there is no
+			// longer a feature gate and issuer URL is not set, the user may not
+			// expect this feature to be enabled. We log the former case as an Error
+			// and the latter case as an Info.
+			klog.Error(msg)
+		} else {
+			klog.Info(msg)
+		}
+	} else {
+		routes.NewOpenIDMetadataServer(md.ConfigJSON, md.PublicKeysetJSON).
+			Install(generic.Handler.GoRestfulContainer)
+	}
+
+	s := &Server{
+		GenericAPIServer: generic,
+
+		APIResourceConfigSource:   c.APIResourceConfigSource,
+		RESTOptionsGetter:         c.Generic.RESTOptionsGetter,
+		ClusterAuthenticationInfo: c.ClusterAuthenticationInfo,
+		VersionedInformers:        c.VersionedInformers,
+	}
+
+	client := kubernetes.NewForConfigOrDie(s.GenericAPIServer.LoopbackClientConfig)
+	if len(c.SystemNamespaces) > 0 {
+		s.GenericAPIServer.AddPostStartHookOrDie("start-system-namespaces-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+			go systemnamespaces.NewController(c.SystemNamespaces, client, s.VersionedInformers.Core().V1().Namespaces()).Run(hookContext.StopCh)
```

Comment:
probably this is the pattern we want to follow, to have "local" properties off commonly-used fields in the structs they are used, but just noting that we could just use `c.VersionedInformers` here, because it's not used anywhere else (for now at least)

similar case for ClusterAuthenticationInfo, it's only used in this setup function, not later, so technically it would not need to be in the struct